File: LinearAlgebra/TypeScript/src/matrices/matrix2.ts
/**
 * @todo Create English Version
 */
// import { BaseMatrix } from "./base_matrix";
/*
class Matrix2 implements BaseMatrix {
  private matrix: Float32Array;
  constructor(matrix: Float32Array = new Float32Array(4)) {
    this.matrix = matrix;
  }
}*/

// 関数の設計
// クラス実装のメソッド


/**
 * Utility Function
 * @description Creating a 2x2 Matrix
 * @static create: (a11: number, a12: number, a21: number, a22: number)
 * @param {number} a11 a11 is a scalar value
 * @param {number} a12 a12 is a scalar value
 * @param {number} a21 a21 is a scalar value
 * @param {number} a22 a22 is a scalar value
 * @returns matrix
 * @example
 * const matrix1 = create(1.0, 0.0, 0.0, 1.0);
 * const matrix2 = create(10.0, 20.0, 30.0, 40.0);
 */
export const create = (a11: number, a12: number, a21: number, a22: number) => {
  const matrix = new Float32Array(4);
  matrix[0] = a11;
  matrix[1] = a12;
  matrix[2] = a21;
  matrix[3] = a22;

  return matrix;
};

/**
 * Utility Function
 * @description Satisfy all components with 0
 * @static init()
 * @returns matrix
 * @example
 * 
 */
export const init = () => {
  const matrix = new Float32Array(4);

  return matrix;
}


/**
 * @description Equality Matrix4
 * @method
 * @param {Float32Array} a Matrix A
 * @param {Float32Array} b Matrix B
 * @returns flag
 */
export const equal = (a: Float32Array, b: Float32Array) => {
  let flag = false;
  let sum = 0;
  if (a.length === b.length) {
    [...Array(a.length).keys()].forEach((count) => {
      if (a[count] === b[count]) {
        sum += 1;
      }
    });
  } else {
    throw new Error("Error Matrix A and B do not have the same size");
  }

  if (sum === 4) {
    flag = true;

    return flag;
  }

  return flag;
};

/**
 * @description Zero Matrix
 * @method
 * @returns matrix
 */
export const zero = () => {
  const matrix = new Float32Array(4);
  matrix[0] = 0;
  matrix[1] = 0;
  matrix[2] = 0;
  matrix[3] = 0;

  return matrix;
};

/**
 * @description Identity Matrix
 * @method
 * @returns matrix
 */
export const identity = () => {
  const matrix = new Float32Array(4);
  matrix[0] = 1;
  matrix[1] = 0;
  matrix[2] = 0;
  matrix[3] = 1;

  return matrix;
};

/**
 * @description matrix addition
 * @param {Float32Array} a Matrix A
 * @param {Float32Array} b Matrix B
 * @returns matrix
 */
export const add = (a: Float32Array, b: Float32Array) => {
  const matrix = new Float32Array(4);
  if (a.length === b.length) {
    [...Array(a.length).keys()].forEach((count) => {
      matrix[count] = a[count] + b[count];
    });
  } else {
    throw new Error("Error Matrix A and B do not have the same size");
  }

  return matrix;
};

/**
 * @description matrix matrix subtraction
 * @param {Float32Array} a Matrix A
 * @param {Float32Array} b Matrix B
 * @returns matrix
 */
export const sub = (a: Float32Array, b: Float32Array) => {
  const matrix = new Float32Array(4);
  if (a.length === b.length) {
    [...Array(a.length).keys()].forEach((count) => {
      matrix[count] = a[count] - b[count];
    });
  } else {
    throw new Error("Error Matrix A and B do not have the same size");
  }

  return matrix;
};

/**
 * @description scalar multiple
 * @param s scalar value
 * @param a matrix A
 * @returns matrix
 */
export const scalarMultiply = (s: number, a: Float32Array) => {
  const matrix = new Float32Array(4);
  if (a.length === 4) {
    [...Array(a.length).keys()].forEach((count) => {
      matrix[count] = s * a[count];
    });
  } else {
    throw new Error("Error The number of matrix components is four.");
  }

  return matrix;
};

/**
 * @description matrix multiply
 * @param a matrix A
 * @param b matrix B
 * @returns matrix
 */
export const multiply = (a: Float32Array, b: Float32Array) => {
  const matrix = new Float32Array(4);
  if (a.length === 4 && b.length === 4) {
    matrix[0] = a[0] * b[0] + a[1] * b[2];
    matrix[1] = a[0] * b[1] + a[1] * b[3];
    matrix[2] = a[2] * b[0] + a[3] * b[2];
    matrix[3] = a[2] * b[1] + a[3] * b[3];
  } else {
    throw new Error(
      "The number of components of matrix A and matrix B is 4 each."
    );
  }
  return matrix;
};

File: LinearAlgebra/Rust/matrices/src/matrix2.rs
pub struct Matrix2 {
  matrix: [f64; 4],
}

impl Matrix2 {
  /// ## コンストラクタ
  /// ```
  /// let mut init_matrix = Matrix2::new(init());
  /// ```
  pub fn new(init_matrix: [f64; 4]) -> Matrix2 {
    Matrix2 {
      matrix: init_matrix,
    }
  }
}

impl Matrix2 {
  /// ## 行列の作成
  /// 2x2行列の作成
  /// ### Example
  /// ```
  /// let mut init_matrix = Matrix2::new(init());
  /// let matrix = init_matrix.create(10.0,20.0,30.0,40.0);
  /// assert_eq!(matrix.len(), 4);
  /// let test_matrix: [f64; 4] = [10.0,20.0,30.0,40.0];
  /// assert_eq!(equal(matrix, test_matrix), true);
  /// ```
  pub fn create(&mut self, a11: f64, a12: f64, a21: f64, a22: f64) -> [f64; 4] {
    self.matrix[0] = a11;
    self.matrix[1] = a12;
    self.matrix[2] = a21;
    self.matrix[3] = a22;

    return self.matrix;
  }

  /// ## 行列の相等
  /// 行列Aと行列Bの対応する成分が等しいかどうか計算する -> 返り値はtureもしくはfalse
  /// ### Example
  /// ```
  /// let mut init_matrix = Matrix2::new(init());
  /// let matrix2: [f64; 4] = [1.0, 0.0, 0.0, 1.0];
  /// let matrix = init_matrix.equal(matrix2);
  /// assert_eq!(matrix, true);
  /// ```
  pub fn equal(&mut self, other: [f64; 4]) -> bool {
    let mut flag = false;
    if self.matrix[0] == other[0]
      && self.matrix[1] == other[1]
      && self.matrix[2] == other[2]
      && self.matrix[3] == other[3]
    {
      flag = true;
      return flag;
    }
    return flag;
  }

  /// ## ゼロ行列
  /// 成分がすべて0の行列
  /// ### Example
  /// ```
  /// let mut init_matrix = Matrix2::new(init());
  /// let matrix = init_matrix.zero();
  /// assert_eq!(matrix.len(), 4);
  /// ```
  pub fn zero(&mut self) -> [f64; 4] {
    self.matrix[0] = 0.0;
    self.matrix[1] = 0.0;
    self.matrix[2] = 0.0;
    self.matrix[3] = 0.0;
    return self.matrix;
  }

  /// ## 行列の足し算
  /// 2x2行列の足し算
  /// ### Example
  /// ```
  /// let matrix1: [f64; 4] = create(10.0, 20.0, 30.0, 40.0);
  /// let matrix2: [f64; 4] = create(10.0, 20.0, 30.0, 40.0);
  /// let mut init_matrix = Matrix2::new(matrix1);
  /// let matrix = init_matrix.add(matrix2);
  /// assert_eq!(matrix.len(), 4);
  /// let test_matrix: [f64; 4] = [20.0, 40.0, 60.0, 80.0];
  /// assert_eq!(equal(matrix, test_matrix), true);
  /// ```
  pub fn add(&mut self, matrix2: [f64; 4]) -> [f64; 4] {
    let mut matrix: [f64; 4] = init();
    matrix[0] = self.matrix[0] + matrix2[0];
    matrix[1] = self.matrix[1] + matrix2[1];
    matrix[2] = self.matrix[2] + matrix2[2];
    matrix[3] = self.matrix[3] + matrix2[3];

    return matrix;
  }

  /// ## 行列の引き算
  /// 2x2行列の引き算
  /// ### Example
  /// ```
  /// let matrix1: [f64; 4] = create(10.0, 20.0, 30.0, 40.0);
  /// let matrix2: [f64; 4] = create(10.0, 20.0, 30.0, 40.0);
  /// let mut init_matrix = Matrix2::new(matrix1);
  /// let matrix = init_matrix.sub(matrix2);
  /// assert_eq!(matrix.len(), 4);
  /// let test_matrix: [f64; 4] = [0.0, 0.0, 0.0, 0.0];
  /// assert_eq!(equal(matrix, test_matrix), true);
  /// ```
  pub fn sub(&mut self, matrix2: [f64; 4]) -> [f64; 4] {
    let mut matrix: [f64; 4] = init();
    matrix[0] = self.matrix[0] - matrix2[0];
    matrix[1] = self.matrix[1] - matrix2[1];
    matrix[2] = self.matrix[2] - matrix2[2];
    matrix[3] = self.matrix[3] - matrix2[3];

    return matrix;
  }
}

/// 初期化用行列
pub fn init() -> [f64; 4] {
  let matrix: [f64; 4] = [1.0, 0.0, 0.0, 1.0];
  return matrix;
}

